src/main/java/picard/illumina/BasecallsConverter.java

```diff
@@ -122,13 +235,161 @@ void setTileLimits(final Integer firstTile, final Integer tileLimit) {
 
     interface ConvertedClusterDataWriter<OUTPUT_RECORD> {
         void write(final OUTPUT_RECORD rec);
-
         void close();
     }
 
-    /**
-     * A comparator for tile numbers, which are not necessarily ordered by the number's value.
-     */
+    private class RecordWriter implements Runnable {
+        private final SortingCollection<CLUSTER_OUTPUT_RECORD> recordCollection;
+        private final ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD> writer;
+        private final String barcode;
+
+        RecordWriter(final ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD> writer,
+                     final SortingCollection<CLUSTER_OUTPUT_RECORD> recordCollection, final String barcode) {
+            this.writer = writer;
+            this.recordCollection = recordCollection;
+            this.barcode = barcode;
+        }
+
+        @Override
+        public void run() {
+            log.debug("Writing out barcode " + barcode);
+            for (final CLUSTER_OUTPUT_RECORD record : recordCollection) {
+                writer.write(record);
+                writeProgressLogger.record(null, 0);
+            }
+        }
+
+        public String getBarcode() {
+            return barcode;
+        }
+    }
+
+    private class Closer implements Runnable {
+        private final ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD> writer;
+        private final String barcode;
+
+        private Closer(final ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD> writer, final String barcode) {
+            this.writer = writer;
+            this.barcode = barcode;
+        }
+
+        @Override
+        public void run() {
+            log.debug("Closing writer for barcode " + barcode);
+            this.writer.close();
+        }
+    }
+
+    private class TileProcessor implements Runnable {
+        private final int tileNum;
+        private final Map<String, SortingCollection<CLUSTER_OUTPUT_RECORD>> barcodeToRecordCollection = new HashMap<>();
```

Comment:
We know the map size ahead of time since we know our total number of barcode writers we need. I suspect the most efficient map is going to be `new Hashmap<>(numBarcodes, 1.0f)`